src/util/src/WaveformUtil.cc

```diff
@@ -0,0 +1,251 @@
+#include <RAT/Log.hh>
+#include <RAT/WaveformUtil.hh>
+
+namespace RAT {
+
+namespace WaveformUtil {
+
+const int INVALID = 9999;
+
+std::vector<double> ADCtoVoltage(const std::vector<UShort_t>& adcWaveform, double voltageRes, double pedestal) {
+  /*
+  Convert a waveform from ADC units to voltage (mV), and optionally subtract a pedestal (given in ADC units)
+  */
+  std::vector<double> voltageWaveform;
+  voltageWaveform.reserve(adcWaveform.size());
+
+  for (UShort_t adcVal : adcWaveform) {
+    voltageWaveform.push_back((adcVal - pedestal) * voltageRes);
+  }
+
+  return voltageWaveform;
+}
+
+double CalculatePedestal(const std::vector<UShort_t>& waveform, int pedWindowLow, int pedWindowHigh) {
+  /*
+  Calculate the baseline (in ADC counts) in the window between low - high samples.
+  */
+  double pedestal = 0;
+
+  if (pedWindowLow > waveform.size()) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of waveform.");
+  } else if (pedWindowLow >= pedWindowHigh) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of pedestal window.");
+  } else if (pedWindowHigh > waveform.size()) {
+    Log::Die("WaveformUtil: End of pedestal window must be at most end of waveform.");
+  }
+
+  // Ensure end of pedestal window is less than waveform size
+  pedWindowHigh = (pedWindowHigh > waveform.size()) ? waveform.size() : pedWindowHigh;
+
+  for (int i = pedWindowLow; i < pedWindowHigh; i++) {
+    pedestal += waveform[i];
+  }
+  pedestal /= (pedWindowHigh - pedWindowLow);
+  return pedestal;
+}
+
+double CalculatePedestal(const std::vector<double>& waveform, int pedWindowLow, int pedWindowHigh) {
+  /*
+  Calculate the baseline (in mV) in the window between low - high samples.
+  */
+  double pedestal = 0;
+
+  if (pedWindowLow > waveform.size()) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of waveform.");
+  } else if (pedWindowLow >= pedWindowHigh) {
+    Log::Die("WaveformUtil: Start of pedestal window must be before end of pedestal window.");
+  } else if (pedWindowHigh > waveform.size()) {
+    Log::Die("WaveformUtil: End of pedestal window must be at most end of waveform.");
+  }
+
+  // Ensure end of pedestal window is less than waveform size
+  pedWindowHigh = (pedWindowHigh > waveform.size()) ? waveform.size() : pedWindowHigh;
+
+  for (int i = pedWindowLow; i < pedWindowHigh; i++) {
+    pedestal += waveform[i];
+  }
+  pedestal /= (pedWindowHigh - pedWindowLow);
+  return pedestal;
+}
+
+std::pair<int, double> FindHighestPeak(const std::vector<double>& voltageWaveform) {
+  /*
+  Calculate the highest peak (in mV) and the corresponding sample.
+  */
+  double voltagePeak = INVALID;
+  int samplePeak = -INVALID;
```

Comment:
No... the negative sign here is still a "Magic number". Should this be necessary, you can have SAMPLE_INVALID, ADC_INVALID, and VOLTAGE_INVALID separately.